zebra/router-id.c

```diff
@@ -155,134 +217,264 @@ void router_id_del_address(struct connected *ifc)
 	struct listnode *node;
 	struct zserv *client;
 	struct zebra_vrf *zvrf = vrf_info_get(ifc->ifp->vrf_id);
+	afi_t afi;
+	struct list *rid_lo;
+	struct list *rid_all;
 
 	if (router_id_bad_address(ifc))
 		return;
 
-	router_id_get(&before, zvrf_id(zvrf));
+	switch (ifc->address->family) {
+	case AF_INET:
+		afi = AFI_IP;
+		rid_lo = zvrf->rid_lo_sorted_list;
+		rid_all = zvrf->rid_all_sorted_list;
+		break;
+	case AF_INET6:
+		afi = AFI_IP6;
+		rid_lo = zvrf->rid6_lo_sorted_list;
+		rid_all = zvrf->rid6_all_sorted_list;
+		break;
+	default:
+		return;
+	}
+
+	router_id_get(afi, &before, zvrf_id(zvrf));
 
 	if (if_is_loopback(ifc->ifp))
-		l = zvrf->rid_lo_sorted_list;
+		l = rid_lo;
 	else
-		l = zvrf->rid_all_sorted_list;
+		l = rid_all;
 
 	if ((c = router_id_find_node(l, ifc)))
 		listnode_delete(l, c);
 
-	router_id_get(&after, zvrf_id(zvrf));
+	router_id_get(afi, &after, zvrf_id(zvrf));
 
 	if (prefix_same(&before, &after))
 		return;
 
 	for (ALL_LIST_ELEMENTS_RO(zrouter.client_list, node, client))
-		zsend_router_id_update(client, &after, zvrf_id(zvrf));
+		zsend_router_id_update(client, afi, &after, zvrf_id(zvrf));
 }
 
 void router_id_write(struct vty *vty)
 {
 	struct vrf *vrf;
 	struct zebra_vrf *zvrf;
 
-	RB_FOREACH (vrf, vrf_name_head, &vrfs_by_name)
-		if ((zvrf = vrf->info) != NULL)
-			if (zvrf->rid_user_assigned.u.prefix4.s_addr
-			    != INADDR_ANY) {
-				if (zvrf_id(zvrf) == VRF_DEFAULT)
-					vty_out(vty, "router-id %s\n",
-						inet_ntoa(
-							zvrf->rid_user_assigned
-								.u.prefix4));
-				else
-					vty_out(vty, "router-id %s vrf %s\n",
-						inet_ntoa(
-							zvrf->rid_user_assigned
-								.u.prefix4),
-						zvrf_name(zvrf));
-			}
+	RB_FOREACH (vrf, vrf_name_head, &vrfs_by_name) {
+		if ((zvrf = vrf->info) == NULL)
+			continue;
+		if (zvrf->rid_user_assigned.u.prefix4.s_addr != INADDR_ANY) {
+			if (zvrf_id(zvrf) == VRF_DEFAULT)
+				vty_out(vty, "ip router-id %pI4\n",
+					&zvrf->rid_user_assigned.u.prefix4);
+			else
+				vty_out(vty, "ip router-id %pI4 vrf %s\n",
+					&zvrf->rid_user_assigned.u.prefix4,
+					zvrf_name(zvrf));
+		}
+		if (!router_id_v6_is_any(&zvrf->rid6_user_assigned)) {
+			if (zvrf_id(zvrf) == VRF_DEFAULT)
+				vty_out(vty, "ipv6 router-id %pI6\n",
+					&zvrf->rid6_user_assigned.u.prefix6);
+			else
+				vty_out(vty, "ipv6 router-id %pI6 vrf %s\n",
+					&zvrf->rid_user_assigned.u.prefix6,
+					zvrf_name(zvrf));
+		}
+	}
 }
 
-DEFUN (router_id,
+DEFUN (ip_router_id,
+       ip_router_id_cmd,
+       "ip router-id A.B.C.D [vrf NAME]",
```

Comment:
vrf NAME should be implied by the vrf subsection we are in.

Reply:
Are you suggesting to remove the optional argument `vrf` from the new commands `ip router-id` and `ipv6 router-id`, and keep it in the old `router-id` one ? Or to break the backward compatibility and remove it from all the router-id commands ?